iroha/src/crypto.rs

```diff
@@ -106,6 +107,66 @@ impl Debug for Signature {
     }
 }
 
+/// Describes an object that can be signed.
+pub trait Sign {
+    /// The bytes that are signed.
+    fn payload(&self) -> Vec<u8>;
+
+    /// Method to get the signature based on the `payload`.
+    fn signature(
+        &self,
+        public_key: &PublicKey,
+        private_key: &PrivateKey,
+    ) -> Result<Signature, String> {
+        Ok(Signature::new(*public_key, &self.payload(), private_key)?)
+    }
+}
+
+/// Container for multiple signatures.
+#[derive(Debug, Clone, Encode, Decode, Default)]
+pub struct Signatures {
+    signatures: BTreeMap<PublicKey, Signature>,
+}
+
+impl Signatures {
+    /// Adds multiple signatures and skips the duplicates.
+    pub fn add_multiple(&mut self, signatures: &[Signature]) {
+        for signature in signatures.iter().cloned() {
+            let _result = self.try_add(signature);
+        }
+    }
+
+    /// Tries to add a signature. If the signature already exists - returns an error.
+    #[allow(clippy::map_entry)]
+    pub fn try_add(&mut self, signature: Signature) -> Result<(), String> {
```

Comment:
It will be great to replace with [Entry API](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry)

Reply:
The thing is we want to hide this logic. The consensus should know if it has added the signature or not, so to decide to send messages or not. It is not needed to manipulate entry manually, it will only complicate consensus code to my mind.

Reply:
It has an idiomatic syntacs entry().or_insert 
Maybe rename the current method and do not return result?

Reply:
I can change the return type to `bool` I guess. But it is easier to work with results in general, and to my mind it makes sense here...

Reply:
No I was talking about removing `try` part and making this method more reliable